framework/src/userobject/ShapeElementUserObject.C

```diff
@@ -0,0 +1,32 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ShapeElementUserObject.h"
+
+template<>
+InputParameters validParams<ShapeElementUserObject>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  return params;
+}
+
+ShapeElementUserObject::ShapeElementUserObject(const InputParameters & parameters) :
+    ElementUserObject(parameters),
+    _phi(_assembly.phi()),
```

Comment:
The difference is that Kernels et. al. have a dedicated `computeJacobian()` that can be called for each coupled variable.  In between each call the shapes are swapped out behind the scenes to make sure that `phi`, etc. are correct for the derivative MOOSE is currently asking for.

See here:

https://github.com/idaholab/moose/blob/devel/framework/src/base/ComputeFullJacobianThread.C#L63

See how the shapes are re-prepared (why are they always preparing? :-) before every invocation of `Kernel::computeJacobian()`?  That makes sure that for each call the `phi` variables contain the correct shapes.

Basically: the right thing to do here is add a `computeJacobian()` call to UserObjects.  That way the Jacobian can be computed for each coupled variable.

Generalizing the Jacobian value storage will be difficult though... still thinking about that.